src/script/sign.h

```diff
@@ -73,6 +76,580 @@ struct SignatureData {
+                    }
+                    s >> redeem_script;
+                    break;
+                }
+                case PSBT_OUT_WITNESSSCRIPT:
+                {
+                    if (!witness_script.empty()) {
+                        throw std::ios_base::failure("Duplicate Key, output witnessScript already provided");
+                    }
+                    s >> witness_script;
+                    break;
+                }
+                case PSBT_OUT_BIP32_DERIVATION:
+                {
+                    uint64_t value_len = ReadCompactSize(s);
+                    // Make sure that the key is the size of pubkey + 1
+                    if (key.size() != CPubKey::PUBLIC_KEY_SIZE + 1 && key.size() != CPubKey::COMPRESSED_PUBLIC_KEY_SIZE + 1) {
+                        throw std::ios_base::failure("Size of key was not the expected size for the type BIP32 keypath");
+                    }
+                    // Read in the pubkey from key
+                    CPubKey pubkey(key.begin() + 1, key.end());
+                    if (!pubkey.IsFullyValid()) {
+                       throw std::ios_base::failure("Invalid pubkey");
+                    }
+                    if (hd_keypaths.count(pubkey) > 0) {
+                        throw std::ios_base::failure("Duplicate Key, output pubkey derivation path already provided");
+                    }
+
+                    // Read in key path
+                    std::vector<uint32_t> keypath;
+                    for (unsigned int i = 0; i < value_len; i += sizeof(uint32_t)) {
+                        uint32_t index;
+                        s >> index;
+                        keypath.push_back(index);
+                    }
+
+                    // Add to map
+                    hd_keypaths.emplace(pubkey, keypath);
+                    break;
+                }
+                // Unknown stuff
+                default: {
+                    if (unknown.count(key) > 0) {
+                        throw std::ios_base::failure("Duplicate Key, key for unknown value already provided");
+                    }
+                    // Read in the value
+                    uint64_t value_len = ReadCompactSize(s);
+                    std::vector<unsigned char> val_bytes;
+                    val_bytes.resize(value_len);
+                    s >> MakeSpan(val_bytes);
+                    unknown.emplace(std::move(key), std::move(val_bytes));
+                    break;
+                }
+            }
+        }
+    }
+
+    template <typename Stream>
+    PSBTOutput(deserialize_type, Stream& s) {
+        Unserialize(s);
+    }
+};
+
+/** A version of CTransaction with the PSBT format*/
+struct PartiallySignedTransaction
+{
+    CMutableTransaction tx;
+    std::vector<PartiallySignedInput> inputs;
+    std::vector<PSBTOutput> outputs;
+    std::map<std::vector<unsigned char>, std::vector<unsigned char>> unknown;
+    uint64_t num_ins = 0;
```

Comment:
Do you still need `num_ins` and `use_in_index`?

Reply:
Removed